smtp/smtpc.c

```diff
@@ -34,10 +34,134 @@
+ *
+ * See ${PREFIX}/include/curl/typecheck-gcc.h
+ *
+ * Undefine those macros to eliminate macro redefinition warnings.
+ */
+#undef curl_easy_getinfo
+#undef curl_easy_setopt
+
+/*
+ * Storage for libcurl function pointers.
+ */
+#define define_func_ptr(func)		\
+	static __typeof__(func) *func##_ptr;
+define_func_ptr(curl_easy_cleanup)
+define_func_ptr(curl_easy_getinfo)
+define_func_ptr(curl_easy_init)
+define_func_ptr(curl_easy_perform)
+define_func_ptr(curl_easy_setopt)
+define_func_ptr(curl_easy_strerror)
+define_func_ptr(curl_slist_append)
+define_func_ptr(curl_slist_free_all)
+#undef define_func_ptr
+
+/* Macros for calling libcurl functions as usual in the code. */
+#define curl_easy_cleanup	curl_easy_cleanup_ptr
+#define curl_easy_getinfo	curl_easy_getinfo_ptr
+#define curl_easy_init		curl_easy_init_ptr
+#define curl_easy_perform	curl_easy_perform_ptr
+#define curl_easy_setopt	curl_easy_setopt_ptr
+#define curl_easy_strerror	curl_easy_strerror_ptr
+#define curl_slist_append	curl_slist_append_ptr
+#define curl_slist_free_all	curl_slist_free_all_ptr
+
+/* dlopen() handle. Saved to call dlclose(). */
+static void *libcurl_handle;
+
+/*
+ * Set <...>_ptr using dlsym() call.
+ *
+ * Return -1 from the function that uses the macro on error (and
+ * set an error into the diagnostics area).
+ */
+#define load_func(libname, dlopen_handle, func) do {				\
+	func##_ptr = dlsym((dlopen_handle), #func);				\
+	if (func##_ptr == NULL) {						\
+		box_error_set(__FILE__, __LINE__, ER_SYSTEM,			\
+			      "Unable to load symbol %s from %s", #func,	\
+			      libname);						\
+		return -1;							\
+	}									\
+} while(0)
+
+int
+smtpc_init(void)
+{
+	int flags = RTLD_NOW | RTLD_LOCAL;
+	/*
+	 * RTLD_DEEPBIND is necessary on Linux and FreeBSD to bind
+	 * libcurl.so dynamic relocations to the same library
+	 * instead of tarantool executable (that also may offer
+	 * those symbols). See the large explcanation in
+	 * smtp/CMakeLists.txt.
+	 *
+	 * dlopen() on Mac OS provides RTLD_DEEPBIND behaviour by
+	 * default and there is no such flag.
+	 *
+	 * Don't know about other OSes.
+	 */
+#ifdef RTLD_DEEPBIND
+	flags |= RTLD_DEEPBIND;
+#endif
+	const char *libname;
+#ifdef __APPLE__
+	libname = "libcurl.dylib";
+#else
+	libname = "libcurl.so";
+#endif
+	libcurl_handle = dlopen(libname, flags);
+	if (libcurl_handle == NULL) {
+		box_error_set(__FILE__, __LINE__, ER_SYSTEM,
+			      "Unable to load %s", libname);
+		return -1;
+	}
+
+	load_func(libname, libcurl_handle, curl_easy_cleanup);
+	load_func(libname, libcurl_handle, curl_easy_getinfo);
+	load_func(libname, libcurl_handle, curl_easy_init);
+	load_func(libname, libcurl_handle, curl_easy_perform);
+	load_func(libname, libcurl_handle, curl_easy_setopt);
+	load_func(libname, libcurl_handle, curl_easy_strerror);
+	load_func(libname, libcurl_handle, curl_slist_append);
+	load_func(libname, libcurl_handle, curl_slist_free_all);
+
+	return 0;
+}
+
+#undef load_func
+
+void
+smtpc_free(void)
```

Comment:
Are you planning to use this function anywhere?

Reply:
Good catch.

Not yet. Maybe in a future. It seems, I should leave a note in the commit message.

We have no any trigger on a module unloading. However I guess it is possible to add a GC handler on the module table. It is also possible to use [on_shutdown trigger](https://github.com/tarantool/tarantool/issues/5723) on recent tarantool versions to free resources before tarantool exit.

Possible use cases are hot update on system's libcurl.so (without restarting of tarantool; looks are quite rare need) and testing with a leak sanitiser (however I don't know whether it tracks dlopen() / dlclose() calls).

Reply:
Added a paragraph into the commit message.

Reply:
However just calling it from a GC handler would not be sufficient: the GC handler may be called after the second initialization call and will set libcurl function pointers to NULL. Since we're not actually experience any problem due to lack of deinitialization, I'll remove it.

I verified that hot reload works:

```diff
diff --git a/test/smtp.test.lua b/test/smtp.test.lua
index f46bcda..eae255f 100755
--- a/test/smtp.test.lua
+++ b/test/smtp.test.lua
@@ -7,6 +7,12 @@ local fiber = require('fiber')
 local socket = require('socket')
 local os = require('os')
 
+package.loaded['smtp'] = nil
+package.loaded['smtp.client'] = nil
+package.loaded['smtp.client.driver'] = nil
+package.loaded['smtp.lib'] = nil
+client = require('smtp').new()
+
 test:plan(1)
 mails = fiber.channel(100)
```

For the record, the description from the previous commit message:

> An attentive reader may find that smtpc_free() is not called from
> anywhere. Sure, we have no any trigger on unloading a module (on 
> `package.loaded.foo = nil`; and the module's table may be saved
> somewhere anyway). However it worth to keep the function up to date with
> smtpc_init() if we'll need it in future to implement hot reload of
> system's libcurl (without tarantool restart) or clean up a leak
> sanitizer results (don't know, however, whether it tracks dlopen() and 
> dlclose()). In theory it is possible to call it at garbage collecting of
> the module table (add a field with cdata with a metatable with a __gc
> metamethod). Or do it on tarantool shutdown using an atexit() trigger,
> tarantool's box.ctl.on_shutdown() or recently added C level
> on_shutdown() (see [3]). Or we can expose a function that a user may 
> call to fully deinitialize the module. Anyway, I don't see a need now,
> so I left it as possible future enhancement.

I replaced it with the following one:

> There is no smtpc_free() function that would complement smtpc_init(),
> because one extra loaded dynamic library is not a problem. I don't want
> to add code that does not solve any real problem. We can add it later if
> it'll be necessary to don't confuse some leak sanitizer.

The patch with the function removal:

```diff
diff --git a/smtp/smtpc.c b/smtp/smtpc.c
index 30e2eb1..f671a4e 100644
--- a/smtp/smtpc.c
+++ b/smtp/smtpc.c
@@ -39,7 +39,7 @@
 
 #include <module.h>
 
-/* {{{ Subsystem (de)initialization */
+/* {{{ Subsystem initialization */
 
 /* https://gcc.gnu.org/onlinedocs/gcc/Alternate-Keywords.html */
 #ifndef __GNUC__
@@ -119,19 +119,6 @@ bind_libcurl_functions(const char *libname, void *libcurl_handle)
 
 #undef load_func
 
-static void
-unbind_libcurl_functions(void)
-{
-	curl_easy_cleanup_ptr = NULL;
-	curl_easy_getinfo_ptr = NULL;
-	curl_easy_init_ptr = NULL;
-	curl_easy_perform_ptr = NULL;
-	curl_easy_setopt_ptr = NULL;
-	curl_easy_strerror_ptr = NULL;
-	curl_slist_append_ptr = NULL;
-	curl_slist_free_all_ptr = NULL;
-}
-
 int
 smtpc_init(void)
 {
@@ -171,18 +158,7 @@ smtpc_init(void)
 	return 0;
 }
 
-void
-smtpc_free(void)
-{
-	unbind_libcurl_functions();
-
-	if (libcurl_handle != NULL) {
-		dlclose(libcurl_handle);
-		libcurl_handle = NULL;
-	}
-}
-
-/* Subsystem (de)initialization }}} */
+/* Subsystem initialization }}} */
 
 int
 smtpc_env_create(struct smtpc_env *env)
diff --git a/smtp/smtpc.h b/smtp/smtpc.h
index d9099e4..a7a0c72 100644
--- a/smtp/smtpc.h
+++ b/smtp/smtpc.h
@@ -35,7 +35,7 @@
 
 #include <curl/curl.h>
 
-/* {{{ Subsystem (de)initialization */
+/* {{{ Subsystem initialization */
 
 /**
  * Initialize the subsystem.
@@ -48,16 +48,7 @@
 int
 smtpc_init(void);
 
-/**
- * Deinitialize the subsystem.
- *
- * Idempotent: may be called before smtpc_init() or after failed
- * initialization.
- */
-void
-smtpc_free(void);
-
-/* Subsystem (de)initialization }}} */
+/* Subsystem initialization }}} */
 
 /** {{{ Environment */
```